openupgrade_scripts/scripts/mass_mailing/15.0.2.5/pre-migration.py

```diff
@@ -0,0 +1,159 @@
+        SET failure_type = CASE failure_type
+            WHEN 'SMTP' THEN 'mail_smtp'
+            WHEN 'RECIPIENT' THEN 'mail_email_invalid'
+            WHEN 'BOUNCE' THEN 'mail_email_invalid'
+            WHEN 'UNKNOWN' THEN 'unknown'
+        END
+        WHERE failure_type IN ('SMTP', 'RECIPIENT', 'BOUNCE', 'UNKNOWN')
+        """,
+    )
+
+
+def _map_mailing_trace_trace_status(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE mailing_trace
+        ADD COLUMN IF NOT EXISTS trace_status varchar
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_trace
+        SET trace_status = CASE state
+            WHEN 'outgoing' THEN 'outgoing'
+            WHEN 'exception' THEN 'error'
+            WHEN 'sent' THEN 'sent'
+            WHEN 'opened' THEN 'open'
+            WHEN 'replied' THEN 'reply'
+            WHEN 'bounced' THEN 'bounce'
+            WHEN 'ignored' THEN 'cancel'
+        END
+        WHERE trace_status IN
+            (
+                'outgoing',
+                'exception',
+                'sent', 'opened',
+                'replied',
+                'bounced',
+                'ignored')
+        """,
+    )
+
+
+def _fill_mailing_mailing_schedule_type(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE mailing_mailing
+        ADD COLUMN IF NOT EXISTS schedule_type varchar
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_mailing
+        SET schedule_type = CASE
+                WHEN  schedule_date IS NOT NULL THEN 'scheduled'
+                ELSE 'now'
+        END
+        """,
+    )
+
+
+def _compute_model_res_id_mailing_trace(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_trace
+        SET model = CASE
+                WHEN im.model != 'mailing.list' THEN im.model
+                ELSE 'mailing.contact'
+                END
+        FROM
+            mailing_trace as mt
+            INNER JOIN mailing_mailing AS mm
+                ON mt.mass_mailing_id = mm.id
+            INNER JOIN ir_model AS im
+                ON im.id = mm.mailing_model_id
+        WHERE mt.model IS NULL
+        """,
+    )
+
+
+def _compute_ab_testing_total_pc(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE utm_campaign
+        ADD COLUMN IF NOT EXISTS ab_testing_total_pc int
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE utm_campaign
+        SET ab_testing_total_pc = mms.sum_pc
+        FROM utm_campaign AS utm
+        INNER JOIN
+            (
+                SELECT campaign_id, SUM(ab_testing_pc) AS sum_pc
+                FROM mailing_mailing AS MM
+                WHERE ab_testing_enabled = True
+                GROUP BY campaign_id
+            ) mms
+            ON mms.campaign_id = utm.id
```

Comment:
@daonamutc1 test giúp em đoạn này có hoạt động anh nhé
```suggestion
        UPDATE utm_campaign uc
        SET ab_testing_total_pc = SELECT_COUNT(*) FROM mailing_mailing AS mm WHERE mm.ab_testing_enabled = True AND uc.id = mm.campaign_id
```